codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,457 @@
+                "pub fn $fnName(input: &[u8]) -> Result<#{Document}, #{Error}>",
+                "Document" to RuntimeType.Document(runtimeConfig),
+                *codegenScope,
+            ) {
+                rustTemplate(
+                    """
+                    let mut tokens_owned = #{json_token_iter}(input).peekable();
+                    let tokens = &mut tokens_owned;
+                    """,
+                    *codegenScope
+                )
+                rustTemplate("let result = #{expect_document}(tokens);", *codegenScope)
+                expectEndOfTokenStream()
+                rust("result")
+            }
+        }
+    }
+
+    private fun RustWriter.expectEndOfTokenStream() {
+        rustBlock("if let Some(_) = tokens.next()") {
+            rustTemplate(
+                "return Err(#{Error}::custom(\"found more JSON tokens after completing parsing\"));",
+                *codegenScope
+            )
+        }
+    }
+
+    private fun RustWriter.deserializeStructInner(shape: StructureShape, includedMembers: Collection<MemberShape>? = null) {
+        val members = includedMembers ?: shape.members()
+        if (members.isNotEmpty()) {
+            objectKeyLoop {
+                rustBlock("match key.as_escaped_str()") {
+                    for (member in members) {
+                        rustBlock("${member.wireName().dq()} =>") {
+                            withBlock("builder = builder.${member.setterName()}(", ");") {
+                                deserializeMember(member, false)
+                            }
+                        }
+                    }
+                    rustTemplate("_ => #{skip_value}(tokens)?", *codegenScope)
+                }
+            }
+        }
+    }
+
+    private fun RustWriter.deserializeMember(memberShape: MemberShape, forceOptional: Boolean) {
+        when (val target = model.expectShape(memberShape.target)) {
+            is StringShape -> deserializeString(target)
+            is BooleanShape -> rustTemplate("#{expect_bool_or_null}(tokens.next())?", *codegenScope)
+            is NumberShape -> deserializeNumber(target)
+            is BlobShape -> rustTemplate("#{expect_blob_or_null}(tokens.next())?", *codegenScope)
+            is TimestampShape -> deserializeTimestamp(memberShape)
+            is CollectionShape -> deserializeCollection(target)
+            is MapShape -> deserializeMap(target)
+            is StructureShape -> deserializeStruct(target)
+            is UnionShape -> deserializeUnion(target)
+            is DocumentShape -> rustTemplate("Some(#{expect_document}(tokens)?)", *codegenScope)
+            else -> TODO(target.toString())
+        }
+        val symbol = symbolProvider.toSymbol(memberShape)
+        if (symbol.isBoxed()) {
+            rust(".map(|v| Box::new(v))")
+        }
+        if (!forceOptional && !symbol.isOptional()) {
+            if (symbol.canUseDefault()) {
+                rust(".unwrap_or_default()")
```

Comment:
you shouldn't need to do this if you use the builder (but I could be wrong).

Reply:
This was the result of having to create union variants that have zero types. For example:

```rust
"int" => Some(crate::model::Choice::Int(
    smithy_json::deserialize::token::expect_number_or_null(tokens.next())?
        .map(|v| v.to_i32())
        .unwrap_or_default(),
)),
```

Reply:
oh right...unions don't have builders. OK.